File: complete_search/46_permutations.py
from itertools import permutations
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:

        if len(nums) == 1:
            return [nums]
        res = []
        *front, last = nums

        for perm in self.permute(front):
            for i in range(len(perm) + 1):
                permutation = perm[:i] + [last] + perm[i:]
                res.append(permutation)
        return res
        # return permutations(nums)

        

File: backtracking/46_permutations.java
class Solution {
    List<List<Integer>> res = new ArrayList<List<Integer>>();
    int[] nums;
    public void backtrack(List<Integer> curr) {
        if (curr.size() == nums.length) {
            res.add(new ArrayList<Integer>(curr));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (curr.contains(nums[i])) continue;
            
            curr.add(nums[i]);
            backtrack(curr);
            curr.remove(curr.size() - 1);
        }
    }
    
    public List<List<Integer>> permute(int[] nums) {
        this.nums = nums;
        
        backtrack(new ArrayList<Integer>());
        
        return res;
    }
}